mitiq/observable/observable.py

```diff
@@ -94,16 +96,21 @@ def matrix(self, dtype: type = np.complex128) -> np.ndarray:
         return matrix
 
     def expectation(
-        self, circuit: QPROGRAM, executor: Callable[[QPROGRAM], QuantumResult]
+        self, circuit: QPROGRAM, execute: Callable[[QPROGRAM], QuantumResult]
     ) -> float:
-        # TODO: Implement. Pass circuits from self.measure_in(circuit=circuit)
-        #  to Collector(executor) to get QuantumResults, then return
-        #  self.expectation_from(quantum_results=quantum_results)
-        raise NotImplementedError
-
-    def expectation_from(self, quantum_results: List[QuantumResult]) -> float:
-        # TODO: Dispatch to correct function based on type of quantum result.
-        raise NotImplementedError
+        result_type = inspect.getfullargspec(execute).annotations.get("return")
+
+        if result_type is MeasurementResult:
+            to_run = self.measure_in(circuit)
+            results = Executor(execute).run(to_run)
+            return self._expectation_from_measurements(
+                cast(List[MeasurementResult], results)
+            )
+        elif result_type is np.ndarray:
+            density_matrix = cast(np.ndarray, execute(circuit))
+            return np.trace(density_matrix @ self.matrix())
+        else:
+            raise NotImplementedError
```

Comment:
Here we could rise a a more explicit error since this can happen often. 
For example: `"The output type of "execute" must be ... or ... but is  ....  ."`

E.g. the user may try the standard old-style executor that returns a float or may just forget annotation.

Reply:
Added.

Reply:
Note that support for batched versions, e.g. `List[mitiq.MeasurementResult]`, need to be added. Meant to add a TODO comment but forgot.